### PR TITLE
Fixing SpanHelpers.LastIndexOfAnyValueType to no longer create out of bounds GC refs

### LINKED ISSUE #75792: GCStress crash stable repro
The following program:
```csharp
public static class Test
{
    private static void Main()
    {
        for (int i = 0; i < 100; i++)
        {
            CreateCollectibleDerived();
            Thread.Sleep(16);
        }
    }

    static void CreateCollectibleDerived()
    {
        ulong collectibleIndex = 0;
        var ab = AssemblyBuilder.DefineDynamicAssembly(
            new AssemblyName($"CollectibleDerivedAssembly{collectibleIndex}"), AssemblyBuilderAccess.Run);
        var mob = ab.DefineDynamicModule($"CollectibleDerivedModule{collectibleIndex}");
        mob.DefineType($"CollectibleDerived{collectibleIndex}", TypeAttributes.Class | TypeAttributes.Public, typeof(Test));
    }
}
```
crashes with
```
Assert failure(PID 42156 [0x0000a4ac], Thread: 23756 [0x5ccc]): SanityCheck()

CORECLR! MethodTable::Validate + 0x4D (0x00007ffe`c04cf77d)
CORECLR! Object::ValidateInner + 0xA9 (0x00007ffe`c04dd8a9)
CORECLR! Object::Validate + 0x88 (0x00007ffe`c04dd7a8)
CORECLR! WKS::GCHeap::Promote + 0xFF (0x00007ffe`c0aadfff)
CORECLR! PromoteCarefully + 0x120 (0x00007ffe`c05537f0)
CORECLR! GcEnumObject + 0x92 (0x00007ffe`c06d73e2)
CORECLR! GcInfoDecoder::ReportRegisterToGC + 0x4F5 (0x00007ffe`c0a2b805)
CORECLR! GcInfoDecoder::ReportSlotToGC + 0xEE (0x00007ffe`c0a2b8fe)
CORECLR! GcInfoDecoder::EnumerateLiveSlots + 0xFD9 (0x00007ffe`c0a2a289)
CORECLR! EECodeManager::EnumGcRefs + 0x5C3 (0x00007ffe`c0390e93)
    File: C:\prj\runtime-main\src\coreclr\vm\methodtable.cpp Line: 8691
    Image: C:\prj\runtime-main\artifacts\bin\coreclr\windows.x64.Checked\corerun.exe
```
with
```
DOTNET_TieredCompilation=0
DOTNET_GCStress=0xC
```
Even on .NET 7.0 branch for me.

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves #75792

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2625]: we don't need `SizeOf` anymore as `offset` is size-independent

```suggestion
            return (int)offset + index;
```

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2633]: same as above

```suggestion
            return (int)offset + index;
```

**Review by adamsitnik — COMMENTED** [review]
@tannergooding big thanks for your help! PTAL at my comments, I think I've found a bug

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2172]: Nit: this is now processing the first vector in the search space regardless of whether any elements remain.  Consider tweaking the comment to remove the condition.

**Review by adamsitnik — APPROVED** [review]
LGTM!

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: I'm being dense, but I'm still not clear on what the actual problem here was.  I see that `currentSearchSpace` could end up pointing to before the object, but at that point it's never dereferenced.  Other than a "SanityCheck" assert, is there an actual negative impact that ref would have?  If it ended up pointing into some unrelated object, it could root that object unnecessarily, and if that object moved, the value of this pointer could end up changing, but if it does change we're not going to use it again, so that shouldn't be visible. Is just the existence of that ref somehow a real problem?

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: The issue is that once it points to before the object, its no longer tracked and won't be updated if the GC decides to relocate `searchSpace`.

This means that if `searchSpace` is moved to a lower address, the comparison will be `false` (and thus loop again) when it should have been `true`.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: > Other than a "SanityCheck" assert, is there an actual negative impact that ref would have?

The GC can move the previous object that the byref pointed into to be on the other side the current buffer that we are searching. It would make the `IsAddressLessThan` check succeed, the loop the continue and eventually either crash or find some completely unrelated match.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: > This means that if searchSpace is moved to a lower address, the comparison will be false (and thus loop again) when it should have been true.

Aha, I see, that makes sense.  It's not that this pointer is a problem in and of itself, it's that the object could be moved to a lower address and this pointer wouldn't be updated, at which point it's pointing past the end of the object.  Got it.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: The same would be true for the inverse direction, where if `currentSearchSpace` pointed past the end of the object and `searchSpace` were moved to a higher address, the comparison would be incorrect.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: Also, the GC is not robust against random byrefs that may be pointing into GC owned memory where no managed reference have business in pointing to.

The contract for byrefs is that they have to be either pointing into unmanaged memory that you own, or they are proper managed byrefs. The details were further clarified in https://github.com/dotnet/runtime/pull/71794 . 

It is not valid to take a random number and convert it to byref. It can crash or corrupt the GC.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: > It is not valid to take a random number and convert it to byref. It can crash or corrupt the GC.

Ok, this is what I thought folks were saying the problem was initially, and I actually didn't realize the GC wasn't robust against that.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: So, does this mean if I have code like:
```C#
byte* ptr = Marshal.AllocHGlobal(...);
ref byte b = ref Unsafe.AsRef<byte>(ptr);
Use(ref b);
Marshal.FreeHGlobal(ptr);
...
```
that if the JIT extends the lifetime of `b` beyond that FreeHGlobal, it's a bug?

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: I think that would depend on the GC implementation.

Most unmanaged heaps are immovable and are isolated such that the GC could never get an address into that range. Thus, it should never view it as a potential managed byref (something pointing in to a managed heap) and will never try to update it.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: > that if the JIT extends the lifetime of b beyond that FreeHGlobal, it's a bug?

Yes, it is a potential bug if the GC allocates the memory block that was just freed and starts using it for managed heap.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: We have a config in GC (`GCConfig::GetConservativeGC`) that enables additional set of validations during stack scanning that are supposed to make the GC robust against the "random" byrefs. It is not heavily tested and probably incomplete. However, we can consider enabling this by default if we think that it is a problem. These checks are relatively cheap (compared to everything else that needs to happen to scan byrefs.)

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: This is an interesting problem. I have opened https://github.com/dotnet/runtime/issues/75865 to have it tracked.

**Comment by tannergooding** [comment]
@stephentoub, would appreciate a secondary sign-off from you before merging.

Great to have a second pair of eyes since we're backporting for .NET 7

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2588]: It doesn't affect this PR, but I wonder why it's written like this rather than:
```C#
if (equals != Vector128<TValue>.Zero)
{
    return ComputeLastIndex(offset, equals);
}
```
Is this just style, or we expect the not found case to be the most common and want the branch predictor to default to that?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs
Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2588]: > we expect the not found case to be the most common and want the branch predictor to default to that

That was my assumption when I was implementing it.

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs
Comment (VSadov) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: I think it would be reasonable to require that the user nulls out the unsafe ref before freeing the memory that it points to.

like:

```cs
byte* ptr = Marshal.AllocHGlobal(...);
ref byte b = ref Unsafe.AsRef<byte>(ptr);
Use(ref b);

// null the ref before releasing the memory
b = ref *(byte*)0;

Marshal.FreeHGlobal(ptr);
```

Once the user started using "Unsafe", they would have to know what they are doing. Especially when representing native memory, not managed by the runtime, as **managed** refs.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: The JIT is free to extend lifetimes and ignore nulling out of locals like this. Also, it is not unusual for the JIT to create invisible temporary locals that are impossible for the user code to access or null out. We would have to prohibit these optimizations in the JIT for byrefs to make the above fully reliable.

Comment (VSadov) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:2170]: But alternative is basically the same as enabling conservative stack scanning. That would break things like unloadability.

**Pull request merged**
